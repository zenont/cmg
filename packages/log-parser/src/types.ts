import type { SensorType } from '@365-widgets/core'
import type { BadLineFormatError } from './error'

export interface GenericSensorSpec {
  readonly threshold: number
}

export interface ThermometerSpec extends GenericSensorSpec {
  readonly deviation: number
}

export type ThermometerSpecConfig = {
  readonly [SensorType.Thermometer]: ThermometerSpec
}

export type HumiditySensorSpecConfig = {
  readonly [SensorType.Humidity]: GenericSensorSpec
}

export type CarbonMonoxideSensorSpecConfig = {
  readonly [SensorType.CarbonMonoxide]: GenericSensorSpec
}

export type SpecConfig =
  | ThermometerSpecConfig
  | HumiditySensorSpecConfig
  | CarbonMonoxideSensorSpecConfig

export type QualityLogResult = Record<string, string>

export interface ReferenceResult {
  /**
   * The control room was held at this constant temperature.
   */
  readonly temperature: number
  /**
   * The control room relative humidity.
   */
  readonly humidity: number
  /**
   * The control room carbon monoxide.
   */
  readonly ppm: number
}

export interface SensorLineResultPayload {
  readonly type: SensorType
  readonly name: string
}

export interface ReferenceLineResult {
  readonly type: 'reference'
  readonly payload: ReferenceResult | BadLineFormatError
}

export interface SensorLineResult {
  readonly type: 'sensor'
  readonly payload: SensorLineResultPayload | BadLineFormatError
}

export interface EntryLineResult {
  readonly type: 'entry'
  readonly payload: number | BadLineFormatError
}

export type LineResult =
  | ReferenceLineResult
  | SensorLineResult
  | EntryLineResult

export interface SensorRead {
  readonly sensorType: SensorType
  readonly max: number
  readonly min: number
  readonly sumValue: number
  readonly totalCount: number
}
