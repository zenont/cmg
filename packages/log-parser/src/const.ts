export const TEMPERATURE_ULTRA_PRECISE_THRESHOLD = 3
export const TEMPERATURE_VERY_PRECISE_THRESHOLD = 5
export const TEMPERATURE_DEGREE_THRESHOLD = 0.5
export const HUMIDITY_THRESHOLD = 1
export const MONOXIDE_PPM_THRESHOLD = 3
