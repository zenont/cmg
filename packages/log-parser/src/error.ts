export class BadLineFormatError extends Error {
  constructor(line: string, lineNumber: number) {
    const formattedLineNumber = Intl.NumberFormat().format(lineNumber)
    super(`Unexpected or invalid line at ${formattedLineNumber}: [${line}]`)
  }
}
