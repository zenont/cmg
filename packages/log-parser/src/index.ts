export * from './types'
export * from './io'
export * from './error'
export * from './util'
export * from './parsers'
export * from './validators'
export * from './evaluator'
