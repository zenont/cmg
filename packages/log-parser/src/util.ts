export const splitLine = (line: string) => line.split(' ').map((x) => x.trim())
