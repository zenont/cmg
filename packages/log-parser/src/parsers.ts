import { isSensorType } from '@365-widgets/core'
import { BadLineFormatError } from './error'
import type {
  ReferenceLineResult,
  SensorLineResult,
  EntryLineResult,
  LineResult,
} from './types'
import { splitLine } from './util'

export const parseReferenceLine = (
  line: string,
  lineNumber: number
): ReferenceLineResult | undefined => {
  const [lineType, ...rest] = splitLine(line)

  if (lineType !== 'reference') {
    return undefined
  }

  const parsedSpec = rest
    .map((x) => Number.parseFloat(x))
    .filter((x) => !Number.isNaN(x))

  const [temperature, humidity, ppm] = parsedSpec
  return {
    type: 'reference',
    payload:
      parsedSpec.length !== 3
        ? new BadLineFormatError(line, lineNumber)
        : {
            temperature,
            humidity,
            ppm,
          },
  }
}

export const parseSensorLine = (
  line: string,
  lineNumber: number
): SensorLineResult | undefined => {
  const [lineType, ...rest] = splitLine(line)

  if (!isSensorType(lineType)) {
    return undefined
  }

  return {
    type: 'sensor',
    payload:
      rest[0] == null || rest[0].trim() === ''
        ? new BadLineFormatError(line, lineNumber)
        : { type: lineType, name: rest[0] },
  }
}

export const parseEntryLine = (
  line: string,
  lineNumber: number
): EntryLineResult | undefined => {
  const [, strValue] = splitLine(line)
  const value = Number.parseFloat(strValue)

  return {
    type: 'entry',
    payload: Number.isNaN(value)
      ? new BadLineFormatError(line, lineNumber)
      : value,
  }
}

/**
 * Parses a line of text from the log file. Validates it in the process to handle malformed or out of spec log files.
 * @param line The log file line.
 * @param lineNumber The log file number.
 * @returns Returns a {@link LineResult} if parsed correctly, or *undefined* if a valid line could not be parsed.
 */
export const parseLine = (
  line: string,
  lineNumber: number
): LineResult | undefined => {
  if (line == null || line.trim() === '') {
    return undefined
  }

  const tuple =
    parseReferenceLine(line, lineNumber) ??
    parseSensorLine(line, lineNumber) ??
    parseEntryLine(line, lineNumber)

  return tuple
}
