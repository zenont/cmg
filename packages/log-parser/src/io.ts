import { Readable, Writable } from 'stream'
import { createInterface } from 'readline'
import type { LineResult } from './types'
import { parseLine } from './parsers'

export type YieldedLineTuple = [LineResult | undefined, number, string]

/**
 * Reads a line at a time from a file. Yields a tuple {@link YieldedLineTuple}
 * @param logContentsStr The file contents as a string.
 */
export async function* readLines(
  logContentsStr: string
): AsyncGenerator<YieldedLineTuple, void, unknown> {
  // we use streams to account for big files
  const readStream = Readable.from(logContentsStr)
  const writeStream = new Writable()
  const reader = createInterface({
    input: readStream,
    output: writeStream,
    terminal: false,
  })

  let lineNumber = 0
  for await (const line of reader) {
    lineNumber++
    const parsedLine = parseLine(line, lineNumber)
    yield [parsedLine, lineNumber, line]
  }
}
