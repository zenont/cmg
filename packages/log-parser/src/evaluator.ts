import { SensorType } from '@365-widgets/core'
import { readLines } from './io'
import { validateSensors } from './validators'
import type { ReferenceResult, QualityLogResult, SensorRead } from './types'

export async function evaluateLogFile(
  logContentsStr: string
): Promise<QualityLogResult> {
  let currReference: ReferenceResult | undefined = undefined
  let currSensorKey: string | undefined = undefined
  let currSensorType: SensorType | undefined = undefined
  const sensors = new Map<string, SensorRead>()
  for await (const [line] of readLines(logContentsStr)) {
    if (line == null) {
      continue
    }

    if (line.payload instanceof Error) {
      throw line.payload
    }

    if (line.type === 'reference') {
      // we scope the reference
      currReference = line.payload
    } else if (line.type === 'sensor') {
      // we scope the sensor
      currSensorKey = line.payload.name
      currSensorType = line.payload.type
    } else if (line.type === 'entry') {
      // we don't include entries if they are not scoped to a sensor
      if (
        currSensorKey == null ||
        currSensorType == null ||
        currReference == null
      ) {
        continue
      }

      const sensor = sensors.get(currSensorKey)
      const sensorType = sensor?.sensorType ?? currSensorType
      const sumValue = (sensor?.sumValue ?? 0) + line.payload
      const totalCount = (sensor?.totalCount ?? 0) + 1
      const min =
        sensor?.min != null ? Math.min(sensor.min, line.payload) : line.payload
      const max =
        sensor?.max != null ? Math.max(sensor.max, line.payload) : line.payload

      sensors.set(currSensorKey, {
        ...sensor,
        sensorType,
        min,
        max,
        sumValue,
        totalCount,
      })
    }
  }

  if (currReference == null) return {}
  return validateSensors(currReference, sensors)
}
