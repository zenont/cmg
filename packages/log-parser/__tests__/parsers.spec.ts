import { parseReferenceLine, parseEntryLine, parseSensorLine } from '../src'

describe('log-parser', () => {
  describe('parsers', () => {
    it('parses the reference line', () => {
      const line = 'reference 70.0 45.0 6'
      const result = parseReferenceLine(line, 1)
      expect(result).toMatchSnapshot()
    })

    it('errors when parsing the reference line', () => {
      const line = 'reference aa 45.0 6'
      const result = parseReferenceLine(line, 1)
      expect(result).toMatchSnapshot()
    })

    it('parses the sensor line', () => {
      const line = 'thermometer temp-1'
      const result = parseSensorLine(line, 1)
      expect(result).toMatchSnapshot()
    })

    it('errors when parsing the sensor line', () => {
      const line = 'thermometer'
      const result = parseSensorLine(line, 1)
      expect(result).toMatchSnapshot()
    })

    it('parses the entry line', () => {
      const line = '2007-04-05T22:00 72.4'
      const result = parseEntryLine(line, 1)
      expect(result).toMatchSnapshot()
    })

    it('errors when parsing the entry line', () => {
      const line = '2007-04-05T22:00 aab'
      const result = parseEntryLine(line, 1)
      expect(result).toMatchSnapshot()
    })
  })
})
