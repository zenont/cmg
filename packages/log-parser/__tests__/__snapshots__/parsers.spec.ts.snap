// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`log-parser parsers errors when parsing the entry line 1`] = `
Object {
  "payload": [Error: Unexpected or invalid line at 1: [2007-04-05T22:00 aab]],
  "type": "entry",
}
`;

exports[`log-parser parsers errors when parsing the reference line 1`] = `
Object {
  "payload": [Error: Unexpected or invalid line at 1: [reference aa 45.0 6]],
  "type": "reference",
}
`;

exports[`log-parser parsers errors when parsing the sensor line 1`] = `
Object {
  "payload": [Error: Unexpected or invalid line at 1: [thermometer]],
  "type": "sensor",
}
`;

exports[`log-parser parsers parses the entry line 1`] = `
Object {
  "payload": 72.4,
  "type": "entry",
}
`;

exports[`log-parser parsers parses the reference line 1`] = `
Object {
  "payload": Object {
    "humidity": 45,
    "ppm": 6,
    "temperature": 70,
  },
  "type": "reference",
}
`;

exports[`log-parser parsers parses the sensor line 1`] = `
Object {
  "payload": Object {
    "name": "temp-1",
    "type": "thermometer",
  },
  "type": "sensor",
}
`;
